source/common/runtime/runtime_impl.cc

```diff
@@ -238,39 +277,65 @@ void SnapshotImpl::walkDirectory(const std::string& path, const std::string& pre
         }
         if (line == lines.back()) {
           const absl::string_view trimmed = StringUtil::rtrim(line);
-          entry.string_value_.append(trimmed.data(), trimmed.size());
+          value.append(trimmed.data(), trimmed.size());
         } else {
-          entry.string_value_.append(std::string{line} + "\n");
+          value.append(std::string{line} + "\n");
         }
       }
-
-      // As a perf optimization, attempt to convert the string into an integer. If we don't
-      // succeed that's fine.
-      uint64_t converted;
-      if (StringUtil::atoul(entry.string_value_.c_str(), converted)) {
-        entry.uint_value_ = converted;
-      }
-
       // Separate erase/insert calls required due to the value type being constant; this prevents
       // the use of the [] operator. Can leverage insert_or_assign in C++17 in the future.
       values_.erase(full_prefix);
-      values_.insert({full_prefix, entry});
+      values_.insert({full_prefix, SnapshotImpl::createEntry(value)});
     }
   }
 }
 
-LoaderImpl::LoaderImpl(Event::Dispatcher& dispatcher, ThreadLocal::SlotAllocator& tls,
-                       const std::string& root_symlink_path, const std::string& subdir,
-                       const std::string& override_dir, Stats::Store& store,
-                       RandomGenerator& generator, Api::OsSysCallsPtr os_sys_calls)
-    : watcher_(dispatcher.createFilesystemWatcher()), tls_(tls.allocateSlot()),
-      generator_(generator), root_path_(root_symlink_path + "/" + subdir),
-      override_path_(root_symlink_path + "/" + override_dir), stats_(generateStats(store)),
+LoaderImpl::LoaderImpl(RandomGenerator& generator, Stats::Store& store,
+                       ThreadLocal::SlotAllocator& tls)
+    : LoaderImpl(DoNotLoadSnapshot{}, generator, store, tls) {
+  loadNewSnapshot();
+}
+
+LoaderImpl::LoaderImpl(DoNotLoadSnapshot /* unused */, RandomGenerator& generator,
+                       Stats::Store& store, ThreadLocal::SlotAllocator& tls)
+    : generator_(generator), stats_(generateStats(store)), admin_layer_(stats_),
+      tls_(tls.allocateSlot()) {}
+
+std::unique_ptr<SnapshotImpl> LoaderImpl::createNewSnapshot() {
+  std::vector<Snapshot::OverrideLayerConstPtr> layers;
+  layers.emplace_back(std::make_unique<const AdminLayer>(admin_layer_));
+  return std::make_unique<SnapshotImpl>(generator_, stats_, std::move(layers));
+}
+
+void LoaderImpl::loadNewSnapshot() {
+  ThreadLocal::ThreadLocalObjectSharedPtr ptr = createNewSnapshot();
+  tls_->set([ptr = std::move(ptr)](Event::Dispatcher&)->ThreadLocal::ThreadLocalObjectSharedPtr {
```

Comment:
nit: You can just capture `ptr` directly without move. With C++14 you might even be able to just do `ptr = createNewSnapshot()`?

Reply:
`std::move` avoids a shared_ptr copy so I did this by instinct. And no, since, we need `ptr` to be the TLS shared_ptr, not a `unique_ptr<Snapshot>`, since the TLS system requires this lambda to be copyable. IMO this is the cleanest way to get the casts/copyability right.

Reply:
OK